Introduction to Deep Leep Learning and Neural Networks with Keras
Week 5 Assignment
Build a Regression Model in Keras

In [1]:
# import required libraries
import pandas as pd # to read csv file into data frame  
import numpy as np  # used for working with arrays

In [2]:
# read csv file from the given link into data frame
concrete_data=pd.read_csv('https://cocl.us/concrete_data/concrete_data.csv')

In [3]:
# show first 5 rows and predictors of the dataset 
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
#concrete_data=concrete_data.drop(['Age'],axis=1)
#concrete_data.head()

In [5]:
concrete_data.shape

(1030, 9)

Data set has 1030 samples and 8 predictors to predict the target value which is concrete strength


Concrete strength is depepndent on Cement, Blast Furnace slag, Fly Ash, Water, Superlasticizer, Coarse Aggregate, Fine Aggregate and Age of cement

Check if the data set has any missing value

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There is no null value/missing value in the given dataset

In [7]:
predictors = concrete_data[['Cement','Blast Furnace Slag','Fly Ash','Water','Superplasticizer','Coarse Aggregate','Fine Aggregate','Age']] # all columns except 'Strngth' are predictors 
target = concrete_data['Strength'] # Target is Strength

In [8]:
# check predictors
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [9]:
[row,col]=predictors.shape
print('Number of columns are: ',col)

Number of columns are:  8


In [10]:
# check target
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [11]:
# import keras for builiding and fitting the model
import keras 
from keras.models import Sequential  # model type
from keras.layers import Dense  # hidden layers

In [12]:
# definition to build regression model
def regression_model():
    # sequential type
    reg_model = Sequential()
    # 10 nodes in hidden layer, number of predictor columns in input layer
    reg_model.add(Dense(10, activation='relu', input_shape=(col,)))
    # one node in output layer 
    reg_model.add(Dense(1))
    
    # compile model
    reg_model.compile(optimizer='adam', loss='mean_squared_error')
    return reg_model

Using library sklearn, rRandomly split dataframe into train and test sets 

In [13]:
from sklearn.model_selection import train_test_split

80% of the datframe is train and 20% is test
Seed of random state is fixed which uses the same split even if this code is executed multiple times  

In [14]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.2, random_state= 10)

In [15]:
# build the model
model = regression_model()

In [16]:
# train model using 50 epochs
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
26/26 [==============================] - 1s 6ms/step - loss: 29971.1426
Epoch 2/50
26/26 [==============================] - 0s 3ms/step - loss: 4193.4155
Epoch 3/50
26/26 [==============================] - 0s 2ms/step - loss: 1844.0699
Epoch 4/50
26/26 [==============================] - 0s 2ms/step - loss: 1420.7780
Epoch 5/50
26/26 [==============================] - 0s 3ms/step - loss: 1134.6667
Epoch 6/50
26/26 [==============================] - 0s 4ms/step - loss: 894.1552
Epoch 7/50
26/26 [==============================] - 0s 6ms/step - loss: 727.4612
Epoch 8/50
26/26 [==============================] - 0s 3ms/step - loss: 605.6348
Epoch 9/50
26/26 [==============================] - 0s 2ms/step - loss: 520.4601
Epoch 10/50
26/26 [==============================] - 0s 3ms/step - loss: 450.6211
Epoch 11/50
26/26 [==============================] - 0s 2ms/step - loss: 407.2305
Epoch 12/50
26/26 [==============================] - 0s 2ms/step - loss: 369.3348
Epoch 13/50
26/26 [

Predict concrete strength using trained model

In [17]:
y_predict = model.predict(X_test)

In [18]:
# evaluate model
# import library to calculate mean square error
from sklearn.metrics import mean_squared_error

In [19]:
mse = mean_squared_error(y_test, y_predict)
mean = np.mean(mse)
stand_dev = np.std(mse)
print('Mean of mean square error: ', mean)
print('Standard Deviation of mean square error: ',stand_dev )

Mean of mean square error:  124.07242787861834
Standard Deviation of mean square error:  0.0


In [20]:
numb_mse = 50
epochs = 50
mean_square_error = []
for i in range(0, numb_mse):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.2, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("Mean Squared Error "+str(i+1)+": "+str(mse))
    y_predict = model.predict(X_test)
    mse_i = mean_squared_error(y_test, y_predict)
    mean_square_error.append(mse_i)

mean_square_error = np.array(mean_square_error)
mean = np.mean(mean_square_error)
stand_dev = np.std(mean_square_error)

print('without normalized data with number of epochs for training is: ', epochs)
print("Mean of Mean Squared error : ", mean)
print("Standard Deviation: ", stand_dev)

Mean Squared Error 1: 104.011962890625
Mean Squared Error 2: 103.3069839477539
Mean Squared Error 3: 66.76934814453125
Mean Squared Error 4: 66.58350372314453
Mean Squared Error 5: 63.37704849243164
Mean Squared Error 6: 60.115211486816406
Mean Squared Error 7: 53.02700424194336
Mean Squared Error 8: 48.52996063232422
Mean Squared Error 9: 47.417362213134766
Mean Squared Error 10: 45.731109619140625
Mean Squared Error 11: 45.721641540527344
Mean Squared Error 12: 42.5257453918457
Mean Squared Error 13: 44.689369201660156
Mean Squared Error 14: 56.30199432373047
Mean Squared Error 15: 46.007205963134766
Mean Squared Error 16: 47.00111770629883
Mean Squared Error 17: 46.38343048095703
Mean Squared Error 18: 53.34968948364258
Mean Squared Error 19: 42.79705810546875
Mean Squared Error 20: 46.41114044189453
Mean Squared Error 21: 41.47123718261719
Mean Squared Error 22: 44.738033294677734
Mean Squared Error 23: 43.0804443359375
Mean Squared Error 24: 49.574039459228516
Mean Squared Error 2